hedera-node/hedera-app/src/main/java/com/hedera/node/app/services/ServiceProviderImpl.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.services;
+
+import com.hedera.node.app.spi.FacilityFacade;
+import com.hedera.node.app.spi.service.Service;
+import com.hedera.node.app.spi.service.ServiceFactory;
+import com.hedera.node.app.spi.service.ServiceProvider;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.ServiceLoader.Provider;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ServiceProviderImpl implements ServiceProvider {
+
+    private final Map<Class<Service>, Service> services;
+
+    public ServiceProviderImpl(final FacilityFacade facilityFacade) {
+        services = new HashMap<>();
+        final ServiceLoader<ServiceFactory> serviceLoader = ServiceLoader.load(ServiceFactory.class);
+        final Set<ServiceFactory> servicesFactories =
+                serviceLoader.stream().map(Provider::get).collect(Collectors.toSet());
+
+        servicesFactories.stream().sorted((a, b) -> sort(a, b)).forEach(factory -> {
+            if (!services.keySet().containsAll(factory.getDependencies())) {
+                throw new IllegalStateException("Can not add service for " + factory.getServiceClass()
+                        + " since the service has a missing dependency");
+            }
+            services.keySet().stream()
+                    .filter(service -> dependOnEachOther(service, factory.getServiceClass()))
+                    .findAny()
+                    .ifPresent(service -> {
+                        throw new IllegalStateException("Can not add service for " + factory.getServiceClass()
+                                + " since the service has a clash with " + service);
+                    });
+            final Service service = factory.createService(this, facilityFacade);
+
+            services.values().stream()
+                    .filter(existingService ->
+                            Objects.equals(existingService.getServiceName(), service.getServiceName()))
+                    .findAny()
+                    .ifPresent(existingService -> {
+                        throw new IllegalStateException("Can not add service for " + service.getClass()
+                                + " since a service with the same name has already been registered.");
+                    });
+
+            services.put(factory.getServiceClass(), service);
+        });
+    }
+
+    /**
+     * Check if two classes are dependent on each other
+     *
+     * @param classA the class to check
+     * @param classB the class to check
+     * @return true if classA is assignable from classB or classB is assignable from classA
+     */
+    private boolean dependOnEachOther(@NonNull final Class<Service> classA, @NonNull final Class<Service> classB) {
+        Objects.requireNonNull(classA, "classA");
```

Comment:
The default message for NullPointerException will tell you the variable name (since Java 14), although see JDK-8233268 for issues with Objects.requireNonNull.
What I would hope to see in a message here is text that explains what went wrong, not simply the name of the variable that's null.  We probably shouldn't always assume that source code is available while reading an exception message.
Example here: "in dependOnEachOther, classA cannot be compared to classB because classA is null".
Yes, it's more typing now, but you'll appreciate that information when reading through diagnostics at 03:00 because you have a high priority bug to fix.

Reply:
Very good point. In Platform we already have a custom solution. sadly platform does not have a solid public API so that we could use it here. will add a better message